### ISSUE TITLE
[Xcode 8.3] pod install don't work

### ISSUE THREAD

**Issue by dral3x** [groue/GRDB.swift#166]
Hi there,

I'm testing our app on iOS10.3 beta so I installed the beta version of Xcode.
I saw you made a specific branch for using GRDB with beta version of Xcode (that's great!) but `pod install` does not succeded because it cannot find the proper `module.modulemap` file.

Here it is the error

```
Errno::ENOENT - No such file or directory @ rb_sysopen - /workspace/ios-radio/Pods/GRDB.swift/Support/module.modulemap
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:1282:in `initialize'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:1282:in `open'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:1282:in `copy_file'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:423:in `copy_file'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:350:in `block in cp'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:1454:in `block in fu_each_src_dest'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:1470:in `fu_each_src_dest0'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:1452:in `fu_each_src_dest'
/usr/local/Cellar/ruby/2.4.0/lib/ruby/2.4.0/fileutils.rb:349:in `cp'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:345:in `block in create_module_map'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:142:in `message'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:344:in `create_module_map'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:27:in `block in install!'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:142:in `message'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:19:in `install!'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator.rb:155:in `block (2 levels) in install_libraries'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator.rb:153:in `each'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator.rb:153:in `block in install_libraries'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:142:in `message'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator.rb:152:in `install_libraries'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/xcode/pods_project_generator.rb:64:in `generate!'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:179:in `block in generate_pods_project'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:64:in `section'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:178:in `generate_pods_project'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:115:in `install!'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/command/install.rb:37:in `run'
/usr/local/lib/ruby/gems/2.4.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/lib/cocoapods/command.rb:50:in `run'
/usr/local/lib/ruby/gems/2.4.0/gems/cocoapods-1.1.1/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:22:in `load'
/usr/local/bin/pod:22:in `<main>'
```

Correctly, inside my project/Pods/GRDB.swift/Support there is no modulemap file.
I see that you made a modulemap for each platform so I'm thinking the problem is in the podspec itself. You need to specify, for each platform, which is the correct modulemap file to load.

Am I right?
Thanks for this great project BTW!

**Comment by groue** [comment]
Hello @dral3x 

The Xcode8.3beta branch is itself very beta 😅 , and I have just made sure the tests were passing. I had to modify the modulemap for that, but I didn't update the podspec.

Your analysis is correct: there are several modulemaps now, [one for each platform](https://github.com/groue/GRDB.swift/tree/627d086a29322bed492320051c03beb6d34fadcd/Support) (macosx, iphones, iphonesimulator, watchOS, watchsimulator). I don't quite know yet how GRDB's podspec should be updated for that - and if those multiple modulemaps are even supported by CocoaPods.

Do you have any insight on these questions?

**Comment by groue** [comment]
BTW the split happened in https://github.com/groue/GRDB.swift/commit/d8c20fb2fcb3e3b37cd37263119cff6eb69fe4ec. Without it, platforms would mess with others (GRDB wouldn't compile for iOS because of the watchOS SDK, and if watchOS was commented out in the modulemap, then OS X would break the iOS framework, etc.)

This may be a bug in Xcode 8.3 beta: I'm not sure. Maybe GRDB will get its single modulemap back when Xcode 8.3 leaves the beta state and is eventually shipped. It's too early to say, and I did not investigate very much - just a little.

On top of that, I'm going to be off for a few weeks, and will be back in business at the end of February: that will be a long time without support.

So... I don't really know what's your best option right now. Help tweaking the podspec? Change the way you integrate GRDB in your app (assuming it would work)? Wait?

**Comment by dral3x** [comment]
Got it!
I tried to just split them by platform, like this
```
s.ios.module_map = 'Support/iphoneos/module.modulemap'
s.osx.module_map = 'Support/macosx/module.modulemap'
s.watchos.module_map = 'Support/watchos/module.modulemap'
```
and at least my project compiles on the device (iOS). Didn't try anything else, so it's not a solution, just a quick-and-dirty workaround for the moment :)

**Comment by groue** [comment]
👍 Looks great! And thanks for the information, it will help :-)

**Comment by dral3x** [comment]
No problem @groue 👍 

I check if I could build for the simulator, and it's not possible in this configuration. It fails while building submodules.

**Comment by groue** [comment]
Yes. CocoaPods doesn't support that today: https://github.com/CocoaPods/CocoaPods/issues/5071

**Comment by dral3x** [comment]
We will keep an eye on that and see how it evolves 👍

**Comment by groue** [comment]
That issue won't evolve much, unless you try to ping the CocoaPods team with some food that tickles their interest - and I think that this requires the issue has to be properly framed, or even [reframed](https://github.com/CocoaPods/CocoaPods/issues/3942). This [comment](https://github.com/CocoaPods/CocoaPods/issues/5071#issuecomment-218517921) by @stephencelis, where he explains how he worked around the issue, may help you.

Meanwhile, remember that GRBD [had](https://github.com/groue/GRDB.swift/commit/d8c20fb2fcb3e3b37cd37263119cff6eb69fe4ec) to split its modulemap because of a strange behavior of Xcode 8.3 beta: nothing says that this problem will persist. Ideally I'd file a radar based on the [previous commit](https://github.com/groue/GRDB.swift/commit/84497bfba7a06f40626ae5cc79313657f1b1917d), where GRDB frameworks would not compile - unfortunately I will lack time until the end of February.

So what I would do, personnaly, is drop CocoaPods for GRDB during the Xcode beta phase. The Carthage approach should fail for the same reasons. Remains the [manual integration option](https://github.com/groue/GRDB.swift#manually) - which you can achieve with a git submodule. You'd call it a day, and start doing your real job: testing your app on iOS 10.3 beta.

**Comment by dral3x** [comment]
I already did my tests on the device, so I'm not blocked by this. 
I will be back on this with future releases of Xcode.

Thanks!

**Comment by groue** [comment]
@dral3x, could you please install XCode8.3beta branch via CocoaPods again?

```ruby
pod 'GRDB.swift', :git => 'https://github.com/groue/GRDB.swift.git', :branch => 'Xcode8.3beta'
```

My last test with Xcode 8.3 beta 4 was successful.

**Comment by dral3x** [comment]
Hi @groue, sorry for this late response.

I downloaded today Xcode 8.3 and I'm working with it (no more "beta").
Fetching GRDB.swift from `Xcode8.3beta` branch seams to work fine.

There is just a single deprecation warning on `Database.swift` file, line 387. No big deal anyway.

Great job!

**Comment by groue** [comment]
Xcode 8.3 users should use the `modulemapRefactor` branch until the next GRDB release.

**Comment by dral3x** [comment]
Good to know. Thanks!

**Comment by groue** [comment]
Superseded by #194

**Comment by groue** [comment]
Support for Xcode 8.3 has shipped in v0.104.0.
